### PR TITLE
Ckdevelop

### PR DESCRIPTION
**Pull request by Roamingturtl**
Implemented Instructions for the following opcodes in emulator.c:

1. 0x05 - DCR B
2. 0x0e - MVI C, D8
3. 0x19 - DAD D
4. 0x26 - MVI H, D8
5. 0x36 - MVI M, D8
6. 0x5e - MOV E, M
7. 0x7a - MOV A, D
8. 0xa7 - ANA A
9. 0xc3 - JMP
10. 0xcd - CALL addr
11. 0xe1 - POP H
12. 0xf1 - POP PSW

Implemented test cases for the following opcodes in tests.c:

1. 0x05 - DCR B
2. 0x0e - MVI C, D8
3. 0x19 - DAD D
4. 0x26 - MVI H, D8
5. 0x36 - MVI M, D8
6. 0x5e - MOV E, M
7. 0x7a - MOV A, D
8. 0xa7 - ANA A
9. 0xc3 - JMP
10. 0xcd - CALL addr
11. 0xe1 - POP H
12. 0xf1 - POP PSW

Can everyone take a look at 0xcd in particular?  Its an important opcode.

### CODE REVIEW

File: emulator.c [hunk 1]
```diff
@@ -81,8 +92,18 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
...
-      printf("DAD D");
-      break;
+      {        // DAD D
+        uint16_t sum
+            = (cpu->h << 8) // NOLINT
+              | (cpu->l);   // convert reg pair h,l to 16 bit int // NOLINT
+        sum = sum + cpu->d;
+
+        cpu->h = sum >> 8;                   // NOLINT
+        cpu->l = sum & 255;                  // NOLINT
```
Comment (flatherz) [inline emulator.c:102]: I think this is similar to what we talked about last night about one of your other instructions. I'm not sure the & is correctly preserving the LSB of sum here...

Comment (cliff-bielinski) [inline emulator.c]: Maybe cpu -> l = (uint8_t) sum & 0x00FF would be clearer?

File: emulator.c [hunk 2]
```diff
@@ -81,8 +92,18 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
...
+        update_carry_flag(cpu, sum > 65535); // NOLINT
```
Comment (cliff-bielinski) [inline emulator.c]: Minor style thing but I think it's clearer to write things in hex. sum > 0xFFFF - easier to be certain you mean the largest 16-bit number.

File: emulator.c [hunk 3]
```diff
@@ -81,8 +92,18 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
         break;
       }
     case 0x19: // NOLINT
-      printf("DAD D");
-      break;
+      {        // DAD D
+        uint16_t sum
+            = (cpu->h << 8) // NOLINT
+              | (cpu->l);   // convert reg pair h,l to 16 bit int // NOLINT
+        sum = sum + cpu->d;
```
Comment (flatherz) [inline emulator.c]: Does it need to add register E as well (or the concatenation of registers D and E)? Not sure, I might be confused here...
image

File: emulator.c [hunk 4]
```diff
@@ -254,9 +308,16 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
         cpu->pc = address;
         return 0;
       }
-    case 0xcd: // NOLINT
-      printf("CALL ");
-      break;
+    case 0xcd:                                            // NOLINT
+      {                                                   // CALL ADDR
+        cpu_write_mem(cpu, cpu->sp - 1, (cpu->pc >> 8));  // NOLINT
+        cpu_write_mem(cpu, cpu->sp - 2, (cpu->pc & 255)); // NOLINT
```
Comment (cliff-bielinski) [inline emulator.c]: Minor style: I would write (cpu->pc & 0xFF) to make clearer it's an 8-bit mask.

File: tests.c [hunk 5]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+
+void
+test_opcode_0x0e(void)
+{
+  // MVI C, D8
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.pc = 0x4341;                       // NOLINT
+  cpu_write_mem(&cpu, cpu.pc + 1, 0x19); // NOLINT
```
Comment (cliff-bielinski) [inline tests.c:187]: Should clean this up so the memory write doesn't persist to the next test.

cpu_write_mem(&cpu, 0x4342, 0x00);

File: emulator.c (hunk 4 above)
Comment (flatherz) [inline emulator.c:314]: Same comment about &ing these lower bits...

File: tests.c [hunk 6]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+void
+test_opcode_0x19(void) // NOLINT
+{
+  // DAD D
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.d = 255; // NOLINT
+  cpu.h = 63;  // NOLINT
+  cpu.l = 200; // NOLINT
```
Comment (cliff-bielinski) [inline tests.c]: Hex would be clearer. Easier to eyeball what a bit-shifted h would look like (0x3F00) compared to decimal.

File: tests.c [hunk 7]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+
+void
+test_opcode_0x26(void)
+{
+  // MVI H, D8
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.pc = 0x4563;                       // NOLINT
+  cpu_write_mem(&cpu, cpu.pc + 1, 0x65); // NOLINT
```
Comment (cliff-bielinski) [inline tests.c:239]: Clean up write memory at end of test.

File: tests.c [hunk 8]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+test_opcode_0x36(void)
+{
+  // MOV M, D8
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.pc = 0x4352;                       // NOLINT
+  cpu.h = 0x23;                          // NOLINT
+  cpu.l = 0x12;                          // NOLINT
+  cpu_write_mem(&cpu, cpu.pc + 1, 0xA1); // NOLINT
```
Comment (cliff-bielinski) [inline tests.c:259]: Clean up write memory at end of test.

File: tests.c [hunk 9]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+void
+test_opcode_0x5e(void)
+{
+  // MOV E, M
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.h = 0x12;                      // NOLINT
+  cpu.l = 0x31;                      // NOLINT
+  cpu_write_mem(&cpu, 0x1231, 0x45); // NOLINT
```
Comment (cliff-bielinski) [inline tests.c:278]: Clean up write memory at end of test.

File: tests.c [hunk 10]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.pc = 0x4341;                       // NOLINT
+  cpu_write_mem(&cpu, cpu.pc + 1, 0x19); // NOLINT
+
+  int code_found = execute_instruction(&cpu, 0x0e); // NOLINT
+
+  CU_ASSERT(code_found == 0);
+  CU_ASSERT_EQUAL(cpu.c, 0x19); // NOLINT
```
Comment (flatherz) [inline tests.c:192]: Don't forget to clean up memory you write too!

File: tests.c [hunk 11]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+void
+test_opcode_0xa7(void) // NOLINT
+{
+  // ANA A
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.a = 85; // NOLINT
+  // S and CY flag set.
+  cpu.flags = 136; // NOLINT
```
Comment (cliff-bielinski) [inline tests.c]: I assume this is all flags set? Maybe clearer to write it out in binary or something?

File: tests.c [hunk 12]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+void
+test_opcode_0xe1(void) // NOLINT
+{
+  // POP H
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.sp = 56030;                       // NOLINT
+  cpu_write_mem(&cpu, cpu.sp, 188);     // NOLINT
+  cpu_write_mem(&cpu, cpu.sp + 1, 254); // NOLINT
```
Comment (cliff-bielinski) [inline tests.c]: Clean up memory writes at end of test

File: tests.c [hunk 13]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+void
+test_opcode_0xf1(void) // NOLINT
+{
+  // POP PSW
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.sp = 0xCBDE;                       // NOLINT
+  cpu_write_mem(&cpu, cpu.sp, 0xA8);     // NOLINT
+  cpu_write_mem(&cpu, cpu.sp + 1, 0xDE); // NOLINT
```
Comment (cliff-bielinski) [inline tests.c:356]: Clean up memory writes at end of test

File: tests.c [hunk 14]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+void
+test_opcode_0xc3(void) // NOLINT
+{
+  // JMP
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.pc = 0xCDEB;                       // NOLINT
+  cpu_write_mem(&cpu, cpu.pc + 1, 0x1E); // NOLINT
+  cpu_write_mem(&cpu, cpu.pc + 2, 0x9c); // NOLINT
```
Comment (cliff-bielinski) [inline tests.c:404]: Clean up memory writes at end of test

File: tests.c [hunk 15]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+{
+  // Call addr
+  i8080 cpu;
+  cpu_init(&cpu);
+
+  cpu.pc = 0xABCD;                       // NOLINT
+  cpu.sp = 0xDCBA;                       // NOLINT
+  uint16_t temp = cpu.sp;                // Holds original sp addr
+  cpu_write_mem(&cpu, cpu.pc + 1, 0xFE); // NOLINT
+  cpu_write_mem(&cpu, cpu.pc + 2, 0xEF); // NOLINT
```
Comment (cliff-bielinski) [inline tests.c:381]: Clean up memory writes at end of test

**Review by cliff-bielinski — CHANGES_REQUESTED** [review]
Overall looks good. For the tests make sure to clean up cpu memory writes cos it persists between unit tests. Also some of the decimal usage was less clear than hex form, and you use hex for some tests/opcodes and decimal for others.  I think maybe all hex would be easier?

File: tests.c [hunk 16]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+
+  int code_found = execute_instruction(&cpu, 0x0e); // NOLINT
+
+  CU_ASSERT(code_found == 0);
+  CU_ASSERT_EQUAL(cpu.c, 0x19); // NOLINT
+}
+
+void
+test_opcode_0x19(void) // NOLINT
+{
```
Comment (flatherz) [inline tests.c:199]: If I'm correct about the missing E register this will need to be updated.

File: tests.c (hunk 7 above)
Comment (flatherz) [inline tests.c:239]: Memory cleanup here too.

File: tests.c [hunk 17]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+  cpu.d = 255; // NOLINT
+  cpu.h = 63;  // NOLINT
+  cpu.l = 200; // NOLINT
+
+  int code_found = execute_instruction(&cpu, 0x19); // NOLINT
+
+  CU_ASSERT(code_found == 0);
+  CU_ASSERT(cpu.h == 64);
+  CU_ASSERT(cpu.l == 199);
+  CU_ASSERT((cpu.flags & FLAG_CY) == 0);
```
Comment (flatherz) [inline tests.c:228]: Should probably also test the case where you DO get a carry.

File: tests.c (hunk 8 above)
Comment (flatherz) [inline tests.c:259]: Cleanup here too.

File: tests.c (hunk 9 above)
Comment (flatherz) [inline tests.c:278]: Cleanup on aisle 257.

File: tests.c [hunk 18]
```diff
@@ -158,6 +159,224 @@ test_opcode_0x13(void) // NOLINT
...
+  cpu_write_mem(&cpu, cpu.sp, 0xA8);     // NOLINT
+  cpu_write_mem(&cpu, cpu.sp + 1, 0xDE); // NOLINT
+  cpu.flags = 0x50;                      // NOLINT
+  cpu.a = 0x19;                          // NOLINT
+
+  int code_found = execute_instruction(&cpu, 0xf1); // NOLINT
+
+  CU_ASSERT(code_found == 0);
+  CU_ASSERT_EQUAL(cpu.flags, 0xA8); // NOLINT
+  CU_ASSERT_EQUAL(cpu.a, 0xDE);     // NOLINT
```
Comment (flatherz) [inline tests.c:364]: More cleanup to do here.

**Review by flatherz — COMMENTED** [review]
See my comment about 0x19. I may be incorrect but I think it might be missing register. Other stuff is cleaning up unit tests and making one more robust.

File: tests.c (hunk 16 above)
Comment (Roamingturtl) [reply tests.c:199]: Yes, you're right register e needed to be included.

**Review by flatherz — APPROVED** [review]
Great work!

**Review by cliff-bielinski — APPROVED** [review]
Nice job!

**Pull request merged**